ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouter.java

```diff
@@ -471,21 +474,73 @@ public Future<String> stitchBlob(BlobProperties blobProperties, byte[] userMetad
 
   /**
    * Requests for a blob to be deleted asynchronously and invokes the {@link Callback} when the request completes.
-   * @param blobId The ID of the blob that needs to be deleted.
-   * @param serviceId The service ID of the service deleting the blob. This can be null if unknown.
-   * @param callback The {@link Callback} which will be invoked on the completion of a request.
+   *
+   * @param restRequest The {@link RestRequest} to delete the blob.
+   * @param blobId      The ID of the blob that needs to be deleted.
+   * @param serviceId   The service ID of the service deleting the blob. This can be null if unknown.
+   * @param callback    The {@link Callback} which will be invoked on the completion of a request.
    * @return A future that would contain information about whether the deletion succeeded or not, eventually.
    */
   @Override
-  public Future<Void> deleteBlob(String blobId, String serviceId, Callback<Void> callback,
+  public Future<Void> deleteBlob(RestRequest restRequest, String blobId, String serviceId, Callback<Void> callback,
       QuotaChargeCallback quotaChargeCallback) {
-    if (blobId == null) {
-      throw new IllegalArgumentException("blobId must not be null");
+    FutureResult<Void> futureResult = new FutureResult<>();
+    if (restRequest == null) {
+      if (blobId == null) {
+        throw new IllegalArgumentException("blobId must not be null");
+      }
+      proceedWithDelete(blobId, serviceId, callback, futureResult, quotaChargeCallback);
+    } else {
+      //if the blobId is not named blob based, it could bypass the first round of d converter logic by checking InternalKeys.BLOB_ID.
+      //First round is to convert named blob to blob Id.
+      if (restRequest.getArgs().get(RestUtils.InternalKeys.BLOB_ID) != null) {
+        String blobIdStr =
+            removeLeadingSlashIfNeeded(restRequest.getArgs().get(RestUtils.InternalKeys.BLOB_ID).toString());
+        proceedWithDelete(blobIdStr, serviceId, callback, futureResult, quotaChargeCallback);
+      } else {
```

Comment:
I am wondering if we can call idConverter.convert() all the time inside the router. For named blobs, it would return the converted blob ID. For regular blobs, it would remove leading slash. Then we wouldn't need if-else condition and `removeLeadingSlashIfNeeded()` method here. What do you think?

Reply:
Logic wise I think it's doable, but I just want to avoid unnecessary network calls to mysql database, that's why I set this internal header to differentiate diff scenarios.

Reply:
Hmm I think in our `IdConverter` logic, we only make calls to MySQL if it is named blobs. For regular blobs, it would just strip the slash

Reply:
Updated.